### DISCUSSION TITLE
Linking mmdeploy to your cpp code

### DISCUSSION THREAD

**Discussion by Likhith-Sugganahalli** [open-mmlab/mmdeploy#1210] [Q&A]
i have gone though all the github issues relating to this, 

according to this, https://github.com/open-mmlab/mmdeploy/issues/280

the op is able to build their project with the cmake that they submitted, but when i do that, the error is 

`CMake Error at /media/likhith/Data/mmdeploy/mmdeploy/build/install/lib/cmake/MMDeploy/modules/FindCUDNN.cmake:18 (message):
  Couldn't find cuDNN in CUDNN_DIR: , or in CUDA_TOOLKIT_ROOT_DIR: , please
  check if the path is correct.
Call Stack (most recent call first):
  /media/likhith/Data/mmdeploy/mmdeploy/build/install/lib/cmake/MMDeploy/MMDeployConfig.cmake:44 (find_package)
  CMakeLists.txt:27 (find_package)`

when i just have 
`find_package(MMDeploy REQUIRED PATHS /media/likhith/Data/mmdeploy/mmdeploy/build/install/lib/cmake/MMDeploy/)`

in my cmake, the error is:
`CMake Error at /media/likhith/Data/mmdeploy/mmdeploy/build/MMDeployConfig.cmake:12 (include):
  include could not find requested file:

    /media/likhith/Data/mmdeploy/mmdeploy/build/MMDeployTargets.cmake
Call Stack (most recent call first):
  CMakeLists.txt:19 (find_package)

CMake Error at /media/likhith/Data/mmdeploy/mmdeploy/build/MMDeployConfig.cmake:43 (find_package):
  By not providing "FindTENSORRT.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "TENSORRT",
  but CMake did not find one.

  Could not find a package configuration file provided by "TENSORRT" with any
  of the following names:

    TENSORRTConfig.cmake
    tensorrt-config.cmake

  Add the installation prefix of "TENSORRT" to CMAKE_PREFIX_PATH or set
  "TENSORRT_DIR" to a directory containing one of the above files.  If
  "TENSORRT" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:19 (find_package)
`

i am not sure how to proceed

**Comment by tehkillerbee** [comment]
@Likhith-Sugganahalli How did you install CUDNN and TensorRT? The first error indicates that CUDNN is not found:
```
Couldn't find cuDNN in CUDNN_DIR: , or in CUDA_TOOLKIT_ROOT_DIR: , please check if the path is correct.
```

Also, I would suggest that you include the full build output and not just snippets.

Can you build the cpp examples included with MMDeploy?

**Comment by Likhith-Sugganahalli** [comment]
@tehkillerbee i am able to build the cpp samples included and run them as well without any issue.

attaching the output of check_env:
![Screenshot from 2022-10-18 11-30-47](https://user-images.githubusercontent.com/44597096/196348050-6eec3b8f-88be-4546-bd01-f53a74c92788.png)

**Comment by tehkillerbee** [comment]
@Likhith-Sugganahalli Please attach the complete (failed) build log when you attempt to build your own CPP application so it is easier to see what goes wrong when the CMake errors occur.

It would also be useful to see how you have written your CMakeList file for your own CPP. It should be similar to the file written for the cpp example codes.

**Reply by Likhith-Sugganahalli** [reply]
@tehkillerbee my bad,

cmake:
![image](https://user-images.githubusercontent.com/44597096/196351506-a69b3b4c-b81c-48d4-81f4-07e7056e6680.png)


output:
![image](https://user-images.githubusercontent.com/44597096/196351219-19a68ad9-766c-45c7-8af1-32b1df312b74.png)

**Reply by Likhith-Sugganahalli** [reply]
@tehkillerbee when i use the same cmake code that builds the code inside the mmdeploy repo, in my package

cmake code:
![image](https://user-images.githubusercontent.com/44597096/196352185-74f2aec7-7dd7-42a5-be5d-c79309ab7a13.png)

cmake output:
![image](https://user-images.githubusercontent.com/44597096/196352246-4a9adce9-4373-4924-a833-44ced1c85357.png)

**Reply by tehkillerbee** [reply]
@Likhith-Sugganahalli In your root CMakeList, you might need to override your TENSORRT DIR and or MMDEPLOY, this was necessary for me when building my own c/cpp inference tool. Something along the lines of the below:

```CMake
# override TENSORRT_DIR, if it is not set
if(NOT ${TENSORRT_DIR})
    set(TENSORRT_DIR "/usr/local/TensorRT-8.2.3.0")
endif()

# Use MMDeploy in current dir if not provided by cmake arguments
if(NOT ${MMDeploy_DIR})
    get_filename_component(MMDeploy_current_path "./MMDeploy/build/install/lib/cmake/MMDeploy" ABSOLUTE)
    set(MMDeploy_DIR ${MMDeploy_current_path})
endif()

# add the subdirectory containing your inference_engine here:
add_subdirectory(inference_engine)
```

It might also be a good idea to check if you have defined TENSORRT_DIR etc in your .bashrc or similar:
```
export PATH="/usr/local/cuda-11.3/bin:$PATH"
export LD_LIBRARY_PATH="/usr/local/cuda-11.3/lib64:$LD_LIBRARY_PATH"
export TENSORRT_DIR="/usr/local/TensorRT-8.2.3.0"
export LD_LIBRARY_PATH="/usr/local/TensorRT-8.2.3.0/lib:$LD_LIBRARY_PATH"
export CUDNN_DIR="/usr/include/x86_64-linux-gnu/"
```

**Note: All the above should be seen as a guideline, replace the respective TensorRT versions, cuda versions etc. with your own installation.**

**Reply by Likhith-Sugganahalli** [reply]
@tehkillerbee i have modified the cmake as per your suggestion

cmake:
![image](https://user-images.githubusercontent.com/44597096/196359294-58a1b502-a5bc-466c-bae6-6b03c3ec3990.png)

i already had the environment variables set as below:
![image](https://user-images.githubusercontent.com/44597096/196359584-1f2db204-0f95-4aa4-9e2f-5ff067e889fb.png)

the error when i build now:
![image](https://user-images.githubusercontent.com/44597096/196360088-0a767c00-f537-494a-8efd-b316e3b3c39b.png)


the MMDeployConfig.cmake file is below:
/media/likhith/Data/mowito/mmdeploy/mmdeploy/build/install/lib/cmake/MMDeploy

**Comment by tehkillerbee** [comment]
@Likhith-Sugganahalli Its odd that your errors are not consistent. Sometimes its an issue about CUDNN, sometimes its TensorRT.

Have you removed your build output directory after each attempt?

It looks like you are working on an external mounted drive? Permission issue, perhaps?

**Reply by Likhith-Sugganahalli** [reply]
@tehkillerbee when i clean the build dir and build using cmake .. the error is the tensorrt one. but when i build it again with the same command, the error is cudnn one

no, even the mmdeploy lib is in the same drive, thats not the issue
